### PR TITLE
submit scheduler module for replication

### PR DESCRIPTION
**Pull request by yuanshuhan**
submit scheduler module for replication

### CODE REVIEW

**Comment by steven-zou** [comment]
@yuanshuhan 

Some suggestions:

* Please add meaningful comments to the commits
* Please squash the commits to reduce the number of commits in the same pr
* Please append DCO by `git commit -s -m "message"` or `go amend -s`
* Please make sure the CI passed (golint is paased)

File: src/replication/ng/scheduler/schedule.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+type Scheduler interface {
+	// Schedule tasks
+	Schedule(srcResources []*model.Resource, destResources []*model.Resource) ([]*model.Task, error)
+	StopTransfer(jobID string) error
+}
+
+// Schedule the task to transfer resouce data
+func (d *DefaultReplicator) Schedule(srcResources []*model.Resource, destResources []*model.Resource) ([]*model.Task, error) {
+	var tasks []*model.Task
+	for _, destResource := range destResources {
```
Comment (ywk253100) [inline src/replication/ng/scheduler/schedule.go]: The srcResource:dstResource is 1:1, not m:n, that means srcResources[0] will be replicated to dstResources[0]

**Comment by steven-zou** [comment]
@yuanshuhan 

Please use the command `gofmt -w -d -e -s $(find . -type d \( -path ./src/vendor -o -path ./tests \) -prune -o -name ‘*.go’ -print)` to fix the `gofmt` error.

File: src/replication/ng/scheduler/schedule.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	return &DefaultReplicator{
+		client: client,
+	}
+}
+
+// Scheduler to schedule the tasks to transfer resource data
+type Scheduler interface {
+	// Schedule tasks
+	Schedule(srcResources []*model.Resource, destResources []*model.Resource) ([]*model.Task, error)
+	StopTransfer(jobID string) error
```
Comment (steven-zou) [inline src/replication/ng/scheduler/schedule.go]: Change `StopTransfer ` to `StopExecution(executionID string)`?

File: src/replication/ng/scheduler/schedule.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+// Schedule the task to transfer resouce data
+func (d *DefaultReplicator) Schedule(srcResources []*model.Resource, destResources []*model.Resource) ([]*model.Task, error) {
+	var tasks []*model.Task
```
Comment (steven-zou) [inline src/replication/ng/scheduler/schedule.go]: Add some guard code to guarantee the parameters are valid?

File: src/replication/ng/scheduler/schedule.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+			}
+
+			newjob.Name = job.ImageTransfer
+			newjob.Parameters = map[string]interface{}{
+				"src_resource": srcResource,
+				"dst_resource": destResource,
+			}
+			uuid, err := d.client.SubmitJob(newjob)
+			if err != nil {
+				log.Errorf("failed to submit the job from %v to %v.err:%s!", srcResource, destResource, err.Error())
```
Comment (ywk253100) [inline src/replication/ng/scheduler/schedule.go]: If got error, the status of the task should be error too?

File: src/replication/ng/scheduler/schedule.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+			}
+			task.JobID = uuid
+			task.Status = ""
```
Comment (steven-zou) [inline src/replication/ng/scheduler/schedule.go]: Can we set to a default status like `pending`?

File: src/replication/ng/scheduler/schedule.go [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+			}
+			task.DstResource = string(dest)
+
+			newjob := &models.JobData{
+				Metadata: &models.JobMetadata{
+					JobKind: job.JobKindGeneric,
+				},
+			}
+
+			newjob.Name = job.ImageTransfer
```
Comment (ywk253100) [inline src/replication/ng/scheduler/schedule.go]: Need `StatusHook` here

File: src/replication/ng/scheduler/schedule.go [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+			task.ResourceType = srcResource.Type
+			task.StartTime = time.Now().UTC()
+			src, err := json.Marshal(srcResource)
+			if err != nil {
+				log.Errorf("failed to marshal the srcResource of %v.err:%s!", srcResource, err.Error())
+				task.Status = "Error"
+				tasks = append(tasks, task)
+				continue
+			}
+			task.SrcResource = string(src)
```
Comment (steven-zou) [inline src/replication/ng/scheduler/schedule.go]: `SrcResource` and `DstResource ` should keep the meaningful full qualified name path with the pattern `<namespace>/<name>:<vtag>`

File: src/replication/ng/scheduler/schedule.go [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+			task.DstResource = string(dest)
+
+			newjob := &models.JobData{
+				Metadata: &models.JobMetadata{
+					JobKind: job.JobKindGeneric,
+				},
+			}
+
+			newjob.Name = job.ImageTransfer
+			newjob.Parameters = map[string]interface{}{
```
Comment (steven-zou) [inline src/replication/ng/scheduler/schedule.go]: Missing webhook endpoint setting

File: src/replication/ng/scheduler/schedule.go (hunk 1 above)
Comment (yuanshuhan) [reply src/replication/ng/scheduler/schedule.go]: what if the two parameters got different length? 
should I throw a exception and quit? 
or should I just transfer the resources as much as what the shorter length is?

File: src/replication/ng/scheduler/schedule.go (hunk 4 above)
Comment (yuanshuhan) [reply src/replication/ng/scheduler/schedule.go]: yes,will mark task status as Error

File: src/replication/ng/scheduler/schedule.go (hunk 2 above)
Comment (yuanshuhan) [reply src/replication/ng/scheduler/schedule.go]: sure

File: src/replication/ng/scheduler/schedule.go (hunk 5 above)
Comment (yuanshuhan) [reply src/replication/ng/scheduler/schedule.go]: as you wish

File: src/replication/ng/scheduler/schedule.go (hunk 7 above)
Comment (yuanshuhan) [reply src/replication/ng/scheduler/schedule.go]: got it

**Comment by yuanshuhan** [comment]
close this pr，because it won't auto sync my latest commit

**Pull request closed without merging**